File: backend/app/schemas/gsc.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class GscStatusOut(BaseModel):
    configured: bool
    connected: bool
    google_email: Optional[str] = None
    site_url: Optional[str] = None
    updated_at: Optional[datetime] = None


class GscOAuthStartOut(BaseModel):
    auth_url: str


class GscSiteOut(BaseModel):
    site_url: str
    permission_level: Optional[str] = None


class GscSiteSelectIn(BaseModel):
    site_url: str = Field(..., min_length=1, max_length=500)


class GscAnalyticsRowOut(BaseModel):
    key: str
    clicks: int
    impressions: int
    ctr: float
    position: float


class GscAnalyticsOut(BaseModel):
    start_date: str
    end_date: str
    site_url: str
    dimension: str
    rows: list[GscAnalyticsRowOut]


class GscSummaryOut(BaseModel):
    start_date: str
    end_date: str
    site_url: str
    clicks: int
    impressions: int
    ctr: float
    position: float

File: frontend/src/api/gsc.ts
import http from './http'

export interface GscStatus {
  configured: boolean
  connected: boolean
  google_email?: string | null
  site_url?: string | null
  updated_at?: string | null
}

export interface GscSite {
  site_url: string
  permission_level?: string | null
}

export interface GscSummary {
  start_date: string
  end_date: string
  site_url: string
  clicks: number
  impressions: number
  ctr: number
  position: number
}

export interface GscAnalyticsRow {
  key: string
  clicks: number
  impressions: number
  ctr: number
  position: number
}

export interface GscAnalytics {
  start_date: string
  end_date: string
  site_url: string
  dimension: string
  rows: GscAnalyticsRow[]
}

export function getGscStatus() {
  return http.get<any, GscStatus>('/gsc/status')
}

export function startGscOAuth() {
  return http.get<any, { auth_url: string }>('/gsc/oauth/start')
}

export function disconnectGsc() {
  return http.delete<any, void>('/gsc/disconnect')
}

export function listGscSites() {
  return http.get<any, GscSite[]>('/gsc/sites')
}

export function selectGscSite(siteUrl: string) {
  return http.put<any, GscStatus>('/gsc/sites/active', { site_url: siteUrl })
}

export function getGscSummary(days = 28) {
  return http.get<any, GscSummary>('/gsc/analytics/summary', { params: { days } })
}

export function getGscAnalytics(params: {
  days?: number
  dimension?: 'query' | 'page'
  row_limit?: number
}) {
  return http.get<any, GscAnalytics>('/gsc/analytics', { params })
}
